kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -257,6 +262,8 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
         check(state is Empty || state is JobNode<*>) // only simple state without lists where children can concurrently add
         check(update !is CompletedExceptionally) // only for normal completion
         if (!_state.compareAndSet(state, update.boxIncomplete())) return false
+        onCancellation(null) // simple state is not a failure
+        onCompletionInternal(update)
```

Comment:
Proposal: move both `onCompletionInternal` and `onCancellation`  into `completeStateFinalization`.
Also, for the sake of simplicity, we can completely remove `onCancellation` and use `onCancelled` instead.
Having both `onCancellation` and `onCancelled` with more or less the same contract looks fragile 
WDYT?

Reply:
We cannot move them to `completeStateFinalization`, because `completeStateFinalization` is invoked _after_ that state is already final, so what happens if we do it in `completeStateFinalization`:

1. Job1's state becomes final.
2. Job1's `onCompletionInternal` is invoked (exception is handled).
3. Waiting coroutines are notified.

So, now if we have another coroutine Job2 doing `Job1.join()` then we have a race -- usually exception in Job1 is handled before the Job2 resumes, but sometimes it could check the state of Job1 in just the right moment (in between steps 1 and 2) and return without suspending when the exception was not handled yet.